pkg/operator/ceph/cluster/cephx.go

```diff
@@ -0,0 +1,421 @@
+// example: client.my-user, []string{"mon", "allow *"} becomes:
+//
+//	[client.my-user]
+//		caps mon = "allow *"
+func genKeyring(clientName, authKey string, clientCaps []string) (string, error) {
+	// DO NOT LOG authKey
+	if authKey == "" {
+		return "", fmt.Errorf("cannot generate %q keyring with empty cephx key", clientName)
+	}
+	o := fmt.Sprintf("[%s]\n", clientName)
+	o += fmt.Sprintf("	key = %s\n", authKey)
+	if len(clientCaps)%2 != 0 {
+		return "", fmt.Errorf("cannot generate %q keyring for caps list %v with uneven number of items", clientName, clientCaps)
+	}
+	for i := 0; i < len(clientCaps); i += 2 {
+		o += fmt.Sprintf("	caps %s = %q\n", clientCaps[i], clientCaps[i+1])
+	}
+	return o, nil
+}
+
+// main routine for rotating the client.admin key
+// should be called after mon daemons are upgraded to current ceph image in the cephcluster spec
+func rotateAdminCephxKey(
+	clusterdCtx *clusterd.Context,
+	clusterInfo *cephclient.ClusterInfo,
+	ownerInfo *k8sutil.OwnerInfo,
+	cephCluster *cephv1.CephCluster,
+) error {
+	if clusterInfo.CephCred.Username != cephclient.AdminUsername {
+		// this shouldn't happen during normal runtime - this could indicate external mode cluster
+		logger.Infof("cannot rotate admin cephx key with non-rotatable username %q for cluster in namespace %q", clusterInfo.CephCred.Username, clusterInfo.Namespace)
+		return nil
+	}
+
+	desiredCephVersion := clusterInfo.CephVersion // TODO: update this when/if WithCephVersionUpdate is implemented
+	shouldRotate, err := keyring.ShouldRotateCephxKeys(
+		cephCluster.Spec.Security.CephX.Daemon, clusterInfo.CephVersion, desiredCephVersion, cephCluster.Status.Cephx.Admin)
+	if err != nil {
+		return errors.Wrap(err, "failed to determine if admin cephx key should be rotated")
+	}
+	if !shouldRotate {
+		logger.Debugf("not rotating admin cephx key for cluster in namespace %q", clusterInfo.Namespace)
+		// no rotation, but still update cephx status - critical for greenfield Uninitialized clusters
+		err := updateCephClusterAdminCephxStatus(clusterdCtx, clusterInfo, false)
+		if err != nil {
+			return errors.Wrap(err, "failed to update admin cephx key status after not rotating")
+		}
+		return nil
+	}
+
+	if err := claimAdminRotationLock(clusterInfo.Namespace); err != nil {
+		return err
+	}
+	defer releaseAdminRotationLock(clusterInfo.Namespace)
+
+	logger.Infof("beginning admin cephx key rotation for cluster in namespace %q", clusterInfo.Namespace)
+
+	// As an optimization, use a (shallow) copy of clusterInfo with a context that won't be canceled
+	// during normal runtime. This disallows the rotation process from being interrupted partway
+	// through in the event of a CephCluster spec change, which can happen for any number of reasons
+	// during normal runtime. While this rotation is designed to be able to be recovered if
+	// interrupted, it is still safest to allow the rotation process to finish in full.
+	clusterInfoCopy := clusterInfo.ShallowCopy()
```

Comment:
Instead of implementing a shallow copy method, would dereferencing the pointer have the desired effect? 
```suggestion
	clusterInfoCopy := *clusterInfo
```

Reply:
Good point. I think it should be the same thing. Especially since all we want to do is change the context